tests/run-make/output-type-permutations/rmake.rs

```diff
@@ -0,0 +1,291 @@
+// In 2014, rustc's output flags were reworked to be a lot more modular.
+// This test uses these output flags in an expansive variety of combinations
+// and syntax styles, checking that compilation is successful and that output
+// files are exactly what is expected, no more, no less.
+// See https://github.com/rust-lang/rust/pull/12020
+
+use run_make_support::{
+    bin_name, cwd, dynamic_lib_name, fs_wrapper, name_not_among, rust_lib_name, rustc,
+    shallow_find_files, static_lib_name,
+};
+
+// Each test takes 4 arguments:
+// `must_exist`: output files which must be found - if any are absent, the test fails
+// `can_exist`: optional output files which will not trigger a failure
+// `dir`: the name of the directory where the test happens
+// `rustc_invocation`: the rustc command being tested
+// Any unexpected output files not listed in `must_exist` or `can_exist` will cause a failure.
+fn assert_expected_output_files(
+    must_exist: &[&'static str],
+    can_exist: &[&'static str],
+    dir: &str,
```

Comment:
Suggestion: so I thought maybe we could have a param struct for `must_exist`, `can_exist` and `dir`, something like

```rs
struct Expectations {
    /// Name of the directory where the test happens.
    test_dir: String,
    /// Output files which must be found. The test fails if any are absent.
    expected_files: Vec<String>,
    /// Allowed output files which will not trigger a failure.
    allowed_files: Vec<String>,
}
```

and callsite (each test call) can be like

```rs
assert_expected_output_files(
    Expectations {
        test_dir: "foo.bar".to_string(),
        expected_files: s!["a.lib", static_lib_name("bar")],
        allowed_files: s!["b.dylib"],
    },
    || { todo!() }
);
```

I find this helpful because it clearly documents which each parameter means local to each `assert_expected_output_files` call.

the `s![]` macro is just a little helper so we can have a mixed list of various kinds of strings (so as long as they impl `ToString`), it just converts them to a simple `Vec<String>`.

```rs
macro_rules! s {
    ( $( $x:expr ),* ) => {
        {
            let mut temp_vec = Vec::new();
            $(
                temp_vec.push($x.to_string());
            )*
            temp_vec
        }
    };
}
```

This avoids having to do the `Box::leak(Box::new(...))` dance you have below.

Reply:
Note that I used a `s![]` declarative macro here and not try to do something like `I: IntoIterator, I::Item: AsRef<str>` because that can't handle heterogeneous types (like having `[&'a str, String]` wouldn't work.

Reply:
Implemented, see latest commit.